performance_test/src/communication_abstractions/ros2_communicator.hpp

```diff
@@ -0,0 +1,186 @@
+  explicit ROS2Communicator(SpinLock & lock)
+  : Communicator(lock),
+    m_node(ResourceManager::get().ros2_node()),
+    m_publisher(nullptr)
+  {
+    m_ROS2QOSAdapter = ROS2QOSAdapter(m_ec.qos()).get();
+  }
+
+  /**
+   * \brief Publishes the provided data.
+   *
+   *  The first time this function is called it also creates the publisher.
+   *  Further it updates all internal counters while running.
+   *
+   * \param data The data to publish.
+   * \param time The time to fill into the data field.
+   */
+  void publish(DataType & data, const std::chrono::nanoseconds time)
+  {
+    if (!m_publisher) {
+      auto ros2QOSAdapter = m_ROS2QOSAdapter;
+      // Workaround for bug where RMW/FastRPS keeps history of volatile publisher.
+      if (ros2QOSAdapter.durability ==
+        rmw_qos_durability_policy_t::RMW_QOS_POLICY_DURABILITY_VOLATILE)
+      {
+        ros2QOSAdapter.history = rmw_qos_history_policy_t::RMW_QOS_POLICY_HISTORY_KEEP_LAST;
+        ros2QOSAdapter.depth = std::size_t(10);  // Setting depth to 10 as a safety net.
+      }
+      // End of workaround.
+
+      m_publisher = m_node->create_publisher<DataType>(
+        Topic::topic_name() + m_ec.pub_topic_postfix(), ros2QOSAdapter);
+
+    }
+    lock();
+    data.time = time.count();
+    data.id = next_sample_id();
+    increment_sent();  // We increment before publishing so we don't have to lock twice.
+    unlock();
+    m_publisher->publish(data);
+  }
+
+  /// Reads received data from ROS 2 using callbacks
+  virtual void update_subscription() = 0;
+
+  /// Returns the accumulated data size in bytes.
+  std::size_t data_received()
+  {
+    return num_received_samples() * sizeof(DataType);
+  }
+
+
+protected:
+  std::shared_ptr<rclcpp::Node> m_node;
+  rmw_qos_profile_t m_ROS2QOSAdapter;
+  /**
+   * \brief Callback handler which handles the received data.
+   *
+   * * Verifies that the data arrived in the right order, chronologically and also consistent with the publishing order.
+   * * Counts recieved and lost samples.
+   * * Calculates the latency of the samples received and updates the statistics accordingly.
+   *
+   * \param data The data received.
+   */
+  void callback(const typename DataType::SharedPtr data)
+  {
+    callback(*data);
+  }
+
+  void callback(DataType& data)
+  {
+    if (m_prev_timestamp >= data.time) {
+      throw std::runtime_error(
+          "Data consistency violated. Received sample with not strictly older timestamp");
+    }
+
+    if (m_ec.roundtrip_mode() == ExperimentConfiguration::RoundTripMode::RELAY) {
+      unlock();
+      publish(data, std::chrono::nanoseconds(data.time));
+      lock();
+    } else {
+      m_prev_timestamp = data.time;
+      update_lost_samples_counter(data.id);
+      add_latency_to_statistics(data.time);
+
+    }
+    increment_received();
+  }
+
+private:
+  std::shared_ptr<::rclcpp::Publisher<DataType, std::allocator<void>>> m_publisher;
+};
+
+
```

Comment:
you can do some clean-up in this file, there are couple of places where you have two unnecessary empty lines

Reply:
Removed the empty lines . Thanks